File: 백준/Silver/11048. 이동하기/이동하기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[][] room = new int[n][m];
        int[][] dp = new int[n + 1][m + 1];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                room[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                dp[i][j] = room[i - 1][j - 1] + Math.max(dp[i][j - 1], Math.max(dp[i - 1][j], dp[i - 1][j - 1]));
            }
        }

        System.out.println(dp[n][m]);

        br.close();


    }

    

}

File: 백준/Silver/11048. 이동하기/이동하기.py
import sys
input = sys.stdin.readline

n,m = map(int,input().split())
room = [list(map(int,input().split()))for _ in range(n)]
dp = [[0]*(m+1) for _ in range(n+1)]

for i in range(1,n+1):
    for j in range(1,m+1):
        dp[i][j] = room[i-1][j-1] + max(dp[i][j-1],dp[i-1][j], dp[i-1][j-1])
print(dp[n][m])
